Sources/StreamChat/Controllers/BackgroundListDatabaseObserver.swift

```diff
@@ -231,15 +235,43 @@ class BackgroundListDatabaseObserver<Item, DTO: NSManagedObject> {
         }
     }
 
-    private func processInitialItems() {
-        processItems(nil)
+    private func getInitialItems() {
+        updateItems(nil, notify: false)
     }
 
-    private func processItems(_ changes: [ListChange<Item>]?) {
+    private func updateItems(_ changes: [ListChange<Item>]?, notify: Bool) {
+        let operation = AsyncOperation { [weak self] _, done in
+            guard let self = self else {
+                done(.continue)
+                return
+            }
+
+            self.frc.managedObjectContext.perform {
+                self.processItems(changes, notify: notify) {
+                    done(.continue)
+                }
+            }
+        }
+
+        processingQueue.addOperation(operation)
+    }
+
+    private func processItems(_ changes: [ListChange<Item>]?, notify: Bool, onCompletion: @escaping () -> Void) {
         mapItems { [weak self] items in
-            self?._items = items
-            let changes = changes ?? items.enumerated().map { .insert($1, index: IndexPath(item: $0, section: 0)) }
-            self?.notifyDidChange(changes: changes)
+            guard let self = self else {
+                onCompletion()
+                return
+            }
+
+            self.queue.async(flags: .barrier) {
+                self._items = items
+            }
+
+            let returnedChanges = changes ?? items.enumerated().map { .insert($1, index: IndexPath(item: $0, section: 0)) }
+            if notify {
+                self.notifyDidChange(changes: returnedChanges)
+            }
+            onCompletion()
```

Comment:
Can we add comments on what is `processItems` + `mapItems` + `getInitialItems` doing, please? 🙏 

Another thing that helps reviewing the code is to follow the top-down approach, where high-level functions are at the top, and low-level at the bottom. It is a bit hard reading the functions right now because we need to constantly scroll again to the top to understand the private low-level helpers.

Reply:
Sure, will add them